File: android/app/src/main/java/app/maptalk/data/PushRepository.kt
package app.maptalk.data

import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore

/**
 * Registers FCM tokens and thread subscriptions for push.
 * Local demo is a no-op.
 */
class PushRepository private constructor(
    private val backend: Backend,
) {

    private sealed interface Backend {
        data class Firestore(
            val db: FirebaseFirestore,
            val currentUid: () -> String?,
        ) : Backend

        data object Local : Backend
    }

    var onError: ((Throwable) -> Unit)? = null

    constructor(
        firestore: FirebaseFirestore,
        currentUid: () -> String?,
    ) : this(Backend.Firestore(firestore, currentUid))

    constructor(local: LocalDemoStore) : this(Backend.Local)

    fun registerDevice(deviceId: String, token: String, platform: String) {
        val firestore = backend as? Backend.Firestore ?: return
        val uid = firestore.currentUid() ?: return
        if (token.isEmpty()) return
        firestore.db.collection(Fs.USERS).document(uid)
            .collection(Fs.DEVICES).document(deviceId)
            .set(
                mapOf(
                    Fs.TOKEN to token,
                    Fs.PLATFORM to platform,
                    Fs.UPDATED_AT to FieldValue.serverTimestamp(),
                ),
            )
            .addOnFailureListener { onError?.invoke(it) }
    }

    fun subscribe(toThreadId: String) {
        val firestore = backend as? Backend.Firestore ?: return
        val uid = firestore.currentUid() ?: return
        if (toThreadId.isEmpty()) return
        firestore.db.collection(Fs.THREADS).document(toThreadId)
            .collection(Fs.SUBSCRIBERS).document(uid)
            .set(mapOf(Fs.SUBSCRIBED_AT to FieldValue.serverTimestamp()))
            .addOnFailureListener { onError?.invoke(it) }
    }
}

File: ios/MapTalk/Data/PushRepository.swift
import FirebaseFirestore
import Foundation

/// Registers FCM tokens and thread subscriptions for push.
@MainActor
final class PushRepository {

    var onError: ((Error) -> Void)?

    private let firestore: Firestore?
    private let currentUid: () -> String?

    init(firestore: Firestore, currentUid: @escaping () -> String?) {
        self.firestore = firestore
        self.currentUid = currentUid
    }

    init(local _: LocalDemoStore) {
        firestore = nil
        currentUid = { nil }
    }

    func registerDevice(deviceId: String, token: String, platform: String) {
        guard let firestore, let uid = currentUid(), !token.isEmpty else { return }
        firestore
            .collection(Fs.users).document(uid)
            .collection(Fs.devices).document(deviceId)
            .setData([
                Fs.token: token,
                Fs.platform: platform,
                Fs.updatedAt: FieldValue.serverTimestamp(),
            ]) { [weak self] error in
                if let error { self?.onError?(error) }
            }
    }

    func subscribe(toThreadId threadId: String) {
        guard let firestore, let uid = currentUid(), !threadId.isEmpty else { return }
        firestore
            .collection(Fs.threads).document(threadId)
            .collection(Fs.subscribers).document(uid)
            .setData([Fs.subscribedAt: FieldValue.serverTimestamp()]) { [weak self] error in
                if let error { self?.onError?(error) }
            }
    }
}
